private/protocol/eventstream/eventstreamapi/signer.go

```diff
@@ -0,0 +1,146 @@
+package eventstreamapi
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/signer/v4"
+	"github.com/aws/aws-sdk-go/private/protocol/eventstream"
+)
+
+const (
+	chunkAlgo            = `AWS4-HMAC-SHA256-PAYLOAD`
+	chunkSignatureHeader = ":chunk-signature"
+	chunkDateHeader      = ":date"
+	timeFormat           = "20060102T150405Z"
+)
+
+type Credentials interface {
+	Get() (credentials.Value, error)
+}
+
+type StreamingChunk interface {
+	Headers() []byte
+	Payload() []byte
+}
+
+type SigningKeyDeriver interface {
+	DeriveSigningKey(date time.Time) ([]byte, error)
+}
+
+type ScopeBuilder interface {
+	BuildScope(date time.Time) (string, error)
+}
+
+type StreamingChunkSigner struct {
+	Region  string
+	Service string
+
+	Credentials Credentials
+
+	prevSig []byte
+}
+
+func NewStreamingChunkSigner(region, service string, credentials Credentials) *StreamingChunkSigner {
+	return &StreamingChunkSigner{
+		Region:      region,
+		Service:     service,
+		Credentials: credentials,
+	}
+}
+
+func (s *StreamingChunkSigner) GetSignature(chunk StreamingChunk, date time.Time) ([]byte, error) {
+	credValue, err := s.Credentials.Get()
+	if err != nil {
+		return nil, err
+	}
+
+	sigKey := v4.DeriveSigningKey(s.Region, s.Service, credValue.SecretAccessKey, date)
+
+	keyPath := v4.BuildSigningScope(s.Region, s.Service, date)
+
+	stringToSign := computeStringToSign(date, keyPath, s.prevSig, chunk.Headers(), chunk.Payload())
+
+	signature := computeHMAC(sigKey, []byte(stringToSign))
+	s.prevSig = signature
+
+	return signature, nil
+}
+
+type ChunkSigner interface {
+	GetSignature(chunk StreamingChunk, date time.Time) ([]byte, error)
+}
+
+type MessageSigner struct {
+	Signer ChunkSigner
+}
+
+func (s MessageSigner) SignMessage(msg *eventstream.Message, date time.Time) error {
```

Comment:
The `MessageSigner` type could stay in the `eventstreamapi` package, but everthing else related to signing could go into the `v4` package. This would remove the need to export the `DeriveSigningKey` and `BuildSigningScope` functions.

This would also decouple the `v4` package from `eventstreamapi`.